### PR TITLE
NER with spacy-transformers and our BioBERT NLI+STS CORD-19

### LINKED ISSUE #319: Train NER model for "PATHWAY" using spaCy
1. Use frozen backbone corresponding to the BioBERT_NLI_STS_CORD19_v1.
2. Use training set = annotations15_EmmanuelleLogette_2020-09-22_raw9_Pathway.jsonl.

**Note**: it's the same of #318 , just using `spacy` 's 2.0 `spacy train` command instead of 🤗's `transformers` interface.

### PR DESCRIPTION
**Pull request by pafonta**
Fixes #319.

## Note

Commits from #248 show in the PR log. This shouldn't. The actual changed files are listed correctly, though.

## Description

Define a `spaCy 3` pipeline for NER using as embedding layer our `Transformer` model `BioBERT NLI+STS CORD-19 v1`.

This is done with only 1 file (`config.cfg`) and it contains the whole pipeline and training parameters in only 140 lines.
Explanations on the config.cfg could be found here: https://github.com/BlueBrain/Search/pull/328#issuecomment-816679707.

Other modified files are DVC files plus a type test in `mining/eval.py`.

## Performances

According to `pipelines/ner/eval.py`:

Without any tuning nor clever choices, the performances are better: `+6.67 % token F1` and +1.59 % entity F1.

| | **entity_f1-score**   | entity_precision   | entity_recall   | **token_f1-score**  | token_precision   | token_recall   |
|-|-----------------|------------------|--------------|------------------|------------------|---------------|
| v0.1.0 | 0.5804 | 0.6852 | 0.5034 | 0.5859 | 0.7117 | 0.4979 |
| PR      | 0.5896 | 0.6529 | 0.5374 | 0.6247 | 0.7167 | 0.5536 |
| delta  | +0.0092 | -0.0323 | +0.0340 | +0.0388 | +0.0050 | +0.0557 |
| %       |  **+1,59** | -4,71 | +6,75 | **+6,62** | +0,70 | +11,19

## How to test?

To get the model and the metrics from the experiment:
```
git checkout 283f6e1
cd data_and_models/metrics/ner/
dvc metrics show biobert_cord19.json
```

To get the metrics from the version v0.1.0 aka "COVID-19 paper":
```
git checkout v0.1.0
cd data_and_models/metrics/ner/
dvc pull pathway.json
cat pathway.json
```

## Checklist

- [x] This PR refers to an issue from the [issue tracker](https://github.com/BlueBrain/Search/issues).
- [x] Documentation and `whatsnew.rst` updated.
- [x] `setup.py` and `requirements.txt` updated with new dependencies.
- [x] All CI tests pass.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -51,10 +51,14 @@ scipy==1.6.1
 scispacy==0.4.0
 sentence-transformers==0.4.1.2
 spacy==3.0.5
+spacy-transformers==1.0.1
+# spaCy 3.0.5 requires cupy-cuda102 < 9.0.0.
+cupy-cuda102==8.6.0
```
Comment (Stannislav) [inline requirements.txt:56]: I tried installing this package on my Mac and I'm getting this:
```text
$ pip install cupy-cuda102==8.6.0
ERROR: Could not find a version that satisfies the requirement cupy-cuda102==8.6.0
ERROR: No matching distribution found for cupy-cuda102==8.6.0
```

File: requirements.txt [hunk 2]
```diff
@@ -51,10 +51,14 @@ scipy==1.6.1
...
 typer==0.3.2
 srsly==2.4.0
 sphinx-bluebrain-theme==0.2.4
 torch==1.7.0
 tox==3.22.0
 tqdm==4.57.0
-transformers==4.3.2
+# spacy-transformers 1.0.1 requires transformers < 4.3.0.
+transformers==4.2.2
```
Comment (Stannislav) [inline requirements.txt:64]: I guess these new dependencies would also need to go into `setup.py` if we decide to incorporate this solution into our main branch. It's probably fine not to bother about it at the moment though.

File: src/bluesearch/mining/eval.py [hunk 3]
```diff
@@ -143,11 +144,8 @@ def spacy2df(
...
             (
-                spacy.pipeline.EntityRuler,
-                spacy.pipeline.EntityRecognizer,
-                spacy.pipeline.Lemmatizer,
-                spacy.pipeline.AttributeRuler,
-                spacy.pipeline.Tagger,
+                spacy.pipeline.ner.EntityRecognizer,
+                spacy_transformers.pipeline_component.Transformer,
             ),
         ):
```
Comment (Stannislav) [inline src/bluesearch/mining/eval.py:150]: Like Francesco mentioned in the zoom - wouldn't this change break the eval script for other models?

File: requirements.txt (hunk 1 above)
Comment (pafonta) [reply requirements.txt:56]: Have you a CUDA-enabled GPU on your Mac?

File: requirements.txt (hunk 2 above)
Comment (pafonta) [reply requirements.txt:64]: That was the subject of the discussion during the stand-up. Having them in `setup.py` is needed for `tox`.

File: src/bluesearch/mining/eval.py (hunk 3 above)
Comment (pafonta) [reply src/bluesearch/mining/eval.py:150]: As mentioned in the PR description, this is a patch and this is only for this experiment. Having a `or` to have this part of the code pass is planned.

Comment (pafonta) [reply src/bluesearch/mining/eval.py]: a3965e79f1a3f489b25fa5209498823ff53ab1e6 moves from a patch to a complete integration.

Tests pass and evaluation works for all models.

File: requirements.txt (hunk 2 above)
Comment (pafonta) [reply requirements.txt:64]: Done in https://github.com/BlueBrain/Search/commit/a3965e79f1a3f489b25fa5209498823ff53ab1e6.

File: requirements.txt (hunk 1 above)
Comment (Stannislav) [reply requirements.txt:57]: As far as I know, there's no CUDA on my Mac.

If this is only a patch for this concrete experiment, then I think it's fine if it only works on systems with a CUDA GPU. Just wanted to point out that if this change makes it into master, then the installation would fail on Macs.

Comment (pafonta) [reply requirements.txt:57]: If there is no CUDA, then that's why it fails. It needs CUDA.

> it only works on systems with a CUDA GPU

`cupy-cuda102` is needed to train the pipeline. Would you have a suggestion on handling this?

**Comment by pafonta** [comment]
UPDATE

Using all **annotations15** as `train` set and **annotations10** + **annotations12** as `dev` set gives:
- a token F1 at `0.6495`,
- which is an increase of `+3.96 %`.

**Comment by pafonta** [comment]
UPDATE

Using `biobert_cord19_v1` (no NLI+STS fine-tuning, added in #333), performances seem `worse`:

| Metric           | Old    | New    | Change    |
|------------------|--------|--------|-----------|
| entity_f1-score  | 0.5896 | 0.5512 | -0.0383  |
| entity_precision | 0.6529 | 0.5735 | -0.0794  |
| entity_recall    | 0.5374 | 0.5306 | -0.0068 |
| token_f1-score   | 0.6247 | 0.5991 | -0.0256   |
| token_precision  | 0.7167 | 0.6303 | -0.0863  |
| token_recall     | 0.5536 | 0.5708 | 0.0172   |

File: requirements.txt (hunk 1 above)
Comment (EmilieDel) [reply requirements.txt:57]: That makes me think, maybe we should create 2 `requirements.txt` files (one for `install_requires` and one for `extra_requires`) to avoid anyone having this kind of problem doing `pip install -r requirements.txt`. What do you think ?

File: data_and_models/pipelines/ner/spacy_transformers/config.cfg [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+[paths]
+train = null
+dev = null
```
Comment (EmilieDel) [inline data_and_models/pipelines/ner/spacy_transformers/config.cfg:22]: This is not related to this PR but I wanted to ask you your opinion @pafonta. I saw that for our "main" models in the ner pipelines, we have a common `config.cfg` for all the models. Do you think it is going to stay like this or it could be we will have several `config.cfg` in the future ? I am sorry, I do not understand quite well all those parameters for now.

File: data_and_models/pipelines/ner/spacy_transformers/dvc.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+stages:
+  train_biobert_cord19:
+    cmd: >-
```
Comment (EmilieDel) [inline data_and_models/pipelines/ner/spacy_transformers/dvc.yaml:3]: I can imagine that this file could be integrated in the main `ner pipeline` if we decide to go for that solution, right ? At least, it is the way it is going to look in the future ?

**Review by EmilieDel — APPROVED** [review]
Thanks for doing those experiments. I am still loss about the `config.cfg` parameters but this is already a point that is going to be addressed!

**Comment by pafonta** [comment]
EXPERIMENT

Note: The performances are not comparable with #329 (different training data, base transformer, hyperparameters, ...).

Freezing the `transformer` component seems to give worse performances.

| Metric           | Old    | New    | Change   |
|------------------|--------|--------|----------|
| entity_f1-score  | 0.5896 | 0.4224 | -0.1671  |
| entity_precision | 0.6529 | 0.4103 | -0.2426  |
| entity_recall    | 0.5374 | 0.4354 | -0.1020   |
| token_f1-score   | 0.6247 | 0.4629 | -0.1618  |
| token_precision  | 0.7167 | 0.4711 | -0.2456  |
| token_recall     | 0.5536 | 0.4549 | -0.0987 |

How to freeze the `transformer` component (`spacy_transformers/config.cfg`):

```
 [components.ner.model.tok2vec]
 @architectures = "spacy-transformers.TransformerListener.v1"
-grad_factor = 1.0
+grad_factor = 0.0
 ```

File: requirements.txt (hunk 1 above)
Comment (pafonta) [reply requirements.txt:57]: > maybe we should create 2 `requirements.txt` files (one for `install_requires` and one for `extra_requires`)

I would support this.

This issue is a part of the bigger discussion reported in #280.

File: data_and_models/pipelines/ner/spacy_transformers/dvc.yaml (hunk 5 above)
Comment (pafonta) [reply data_and_models/pipelines/ner/spacy_transformers/dvc.yaml:3]: Yes :)

I have done it like this because that's not clear to me if this PR is more than an experiment or not at the moment. Indeed, it seems the work continues on #329.

File: data_and_models/pipelines/ner/spacy_transformers/config.cfg (hunk 4 above)
Comment (pafonta) [reply data_and_models/pipelines/ner/spacy_transformers/config.cfg:22]: > we have a common `config.cfg` for all the models. Do you think it is going to stay like this

Yes.

> wanted to ask you your opinion

My opinion is that we could integrate `DVC` + `spaCy` in the following way:

1. We use `params.yaml` from `DVC` to overwrite parameters from the default `config.cfg` from `spaCy`.
2. We have a `Python` script which creates a `config.cfg` with the parameters from `params.yaml` overwritten.
3. This step is a DVC stage.
4. We could use `dvc exp run --set-params`, `dvc run --params`, and `dvc params diff` transparently 🚀 

What do you think?

Comment (EmilieDel) [reply data_and_models/pipelines/ner/spacy_transformers/config.cfg:22]: I see! I did not think of it that way but it is really interesting! Thanks for the answer!

**Comment by pafonta** [comment]
Hello @EmilieDel !

Thank you for your review.

> I am still lost about the `config.cfg` parameters but this is already a point that is going to be addressed!

Maybe this video (between 05:03​ and 11:06​) could give some explanations on _why_ `config.cfg` exists and `how` it works:
https://www.youtube.com/watch?v=9k_EfV7Cns0&t=303s

Regarding the sections and their individual parameters in `config.cfg`, the explanations could be found in the documentation, with https://spacy.io/api/data-formats#config as a starting point. Otherwise, one could use the search feature on https://spacy.io (e.g. searching for `spacy-transformers.TransformerModel.v1`).

I have listed below the links linked from this documentation page for each of the parts of the `config.cfg` of the PR. I hope it would help navigate the documentation.

This could seem at first overwhelming but at the end this is the contrary, as explained in the video.

`[paths]` and `[system]`
https://spacy.io/api/data-formats#config-variables

`[nlp]` and `[nlp.tokenizer]`
https://spacy.io/api/data-formats#config-nlp

`[components]`
https://spacy.io/api/data-formats#config-components

`[components.ner]`
https://spacy.io/api/entityrecognizer#config

`[components.ner.model]`
https://spacy.io/api/architectures#TransitionBasedParser

`[components.ner.model.tok2vec]`
https://spacy.io/api/architectures#TransformerListener

`[components.ner.model.tok2vec.pooling]`
https://thinc.ai/docs/api-layers#reduce_mean

`[components.transformer]`
https://spacy.io/api/transformer#config

`[components.transformer.set_extra_annotations]`
https://spacy.io/api/transformer#annotation_setters

`[components.transformer.model]`
https://spacy.io/api/architectures#TransformerModel

`[components.transformer.model.get_spans]`
https://spacy.io/api/transformer#strided_spans

`[components.transformer.model.tokenizer_config]`
https://huggingface.co/transformers/model_doc/auto.html#transformers.AutoTokenizer.from_pretrained

`[corpora]`
https://spacy.io/api/data-formats#config-corpora

`[corpora.train]` and `[corpora.dev]`
https://spacy.io/api/corpus#config

`[training]`
https://spacy.io/api/data-formats#config-training
`
[training.batcher]`
https://spacy.io/api/top-level#batch_by_padded

`[training.logger]`
https://spacy.io/api/top-level#ConsoleLogger

`[training.optimizer]`
https://thinc.ai/docs/api-optimizers#adam

`[training.optimizer.learn_rate]`
https://thinc.ai/docs/api-schedules#warmup_linear

`[training.score_weights]`
https://spacy.io/usage/training#metrics

`[initialize]`
https://spacy.io/api/data-formats#config-initialize

**Comment by EmilieDel** [comment]
Thanks a lot for the precious information @pafonta! I will definitely check this!

**Comment by pafonta** [comment]
EXPERIMENT

Variation on entity F1 (from `eval.py`) compared to BBS v0.1.0 with the `transformer` component frozen or not:

frozen or not - 9 entities

**Comment by FrancescoCasalegno** [comment]
### Experiment
(thanks @pafonta)

What happens if we have more than 1 `ner` component in our `spaCy` pipeline and they happen to predict spans that have non-null intersection?

E.g. "I like to eat potato salad."
`span_1="potato salad"`, `span_2="potato salad"`
`span_1="potato"`, `span_2="potato salad"`
`span_1="potato salad"`, `span_2="salad"`
`span_1="eat potato"`, `span_2="potato salad"`

```
# # CASE 1 | span_1="potato salad", span_2="potato salad"
# # Result:
# # 4 6 FOOD_2 potato salad
```
```
# # CASE 2 | span_1="potato", span_2="potato salad"
# # Result:
# # 4 6 FOOD_2 potato salad
```

```
# # CASE 3 | span_1="potato salad", span_2="salad"
# # Result:
# # 4 5 FOOD_1 potato
# # 5 6 FOOD_2 salad
```

```
# # CASE 4 | span_1="eat potato", span_2="potato salad"
# # Result:
# # 3 4 FOOD_1 eat
# # 4 6 FOOD_2 potato salad
```

**Comment by FrancescoCasalegno** [comment]
The results seem quite promising, so let's try to apply this idea to our production NER models in `data_and_models/pipelines/ner`. 

The conclusions of our discussion as well as the next steps needed in this direction are found in #335.

**Comment by pafonta** [comment]
EXPERIMENT

Last of the discussed experiment.

## Results

Entity F1 compared to BBS v0.1.0 with the transformer from `BioBERT NLI+STS CORD-19 v1` or `BioBERT v1.1`:

bioberts comparison

## Notes

The entity F1 score is from `ner/eval.py`.

As for BBS v0.1.0, most models handle `several` entities (see the row sub-divisions in the table above).

The column `BioBERT (ours)` corresponds to our model trained with [sentence_embedding/scripts/build.sh](https://github.com/BlueBrain/Search/blob/master/data_and_models/pipelines/sentence_embedding/scripts/build.sh).

The column `BioBERT` corresponds to the original model [BioBERT v1.1](https://huggingface.co/dmis-lab/biobert-v1.1/tree/main) from the DMIS Lab.

## Implementation

The change in `config.cfg`:

```
 [components.transformer.model]
 @architectures = "spacy-transformers.TransformerModel.v1"
-name = "../../models/sentence_embedding/biobert_nli_sts_cord19_v1/0_Transformer/"
+name = "dmis-lab/biobert-v1.1"
```
